spec/src/main/asciidoc/changelog.adoc

```diff
@@ -57,6 +57,7 @@ Changes marked with icon:bolt[role="red"] are breaking changes relative to previ
 *** Changed `Timer.update(long duration, java.util.concurrent.TimeUnit)` to `Timer.update(java.time.Duration duration)`
 *** Added `Timer.getElapsedTime()` which returns `java.time.Duration`
 ** Removed `MetadataBuilder.withOptional*` methods, the remaining `with*` methods do accept `null` value (considered not present) except `withName` which does not accept `null` or `""`
+** Global tags and `_app` tag are no longer handled automatically by the `MetricID` class, the implementation is expected to add them by itself, for example during metric export
```

Comment:
`_app` tag differentiates instances of metrics so that different apps can have the same metric name tracked separately in different apps.  So it can't be added at export time -- it needs to happen at or before metric registration time.

Reply:
Not sure I understand this.

If it's separate apps, and the tag is used within Prometheus, for instance, to distinguish between apps, why can't it be added as part of the export?

Reply:
Imagine I have two apps, app A and app B, and they both have a counter named "counterX".  
Imagine now that, within app A, I increment the counter 5 times, and within app B I increment the counter 10 times.

Under the current approach to attaching the _app tag pre-registration I would see the following in my exported metrics:
```
counterX{_app=A} 5
counterX{_app=B} 10
```

Under the proposed approach of attaching the _app at export time I would see:
```
counterX{_app=A} 15
counterX{_app=B} 15
```

Reply:
So you're talking two apps in a single instance right?

From what you were saying before, it didn't seem to indicate that

Reply:
> So you're talking two apps in a single instance right?

yes, exactly.

Reply:
Is it not possible to define the spec in a way that doesn't require the `_app` tag to be added on metric registration, but doesn't preclude a multi instance runtime from doing so?

My concern is specific language and functionality that is only applicable to a multi instance runtime being baked into the spec. What we should be doing is allowing multi instance runtimes to do what they need, but without defining what that is in the spec if it means a single instance runtime has to operate in a sub optimal way

Reply:
I think it is left to the impl. e.g. the implementation of an Interceptor can add that tag (as the required context would be available). I also think the presence/absence of that tag must remain optional (as the spec says it is).

IFF an impl supports multiple applications, THEN the impl can decide how/when to add that tag.

Reply:
Right, the `_app` tag has always been optional.  See https://github.com/eclipse/microprofile-metrics/blob/master/spec/src/main/asciidoc/architecture.adoc#usage-of-microprofile-metrics-in-application-servers-with-multiple-applications which says... 

> It is allowed for application servers to choose to not add the _app tag at all, but in that case, metrics from two applications on one server can clash as no differentiator (by application) is given.

Reply:
I guess my question is, what is wrong with the wording proposed?

It just uses export as an example, but an impl is free to attach it whenever they choose, right?

Reply:
In Payara we use multiple registries, one per app but even if we would use one registry for all apps there would always be the moment when the app context had to be looked up to either pick the right registry or the right metric in a common registry. This doesn't even need the app tag at all if internally the metrics are organised accordingly and if all request on the registry always take the app context as an additional "key" into account. This key could then be made visible as the app tag but the only place were it really needs to be is the export as app context can no longer be derived from the processed request or thread.

Reply:
I'd say the current proposed wording lets implementations attach the `_app` tag whenever they want, so if there is a need to do it during registration, then so be it. 
For annotated metrics, this shouldn't be a problem because the annotation scanning logic can just append the tag
For programmatically registered metrics, this should be possible too - the `MetricRegistry` impl can for example use the TCCL to determine which application the metric (that is being registered) should belong to

Reply:
There's another more specific wording in this PR that says tags must be added during export.  I'll put a comment on that as I think that's not in the spirit of letting impls add the tags whenever they want (eg. pre-registration).

Reply:
_global_ tags at export. don't cross the streams. :-P